pkgs/os-specific/darwin/apple-source-releases/mkAppleDerivation.nix

```diff
@@ -0,0 +1,90 @@
+{
+  lib,
+  bootstrapStdenv,
+  fetchFromGitHub,
+  meson,
+  ninja,
+  stdenv,
+  stdenvNoCC,
+  xcodeProjectCheckHook,
+}:
+
+let
+  hasBasenamePrefix = prefix: file: lib.hasPrefix prefix (baseNameOf file);
+in
+lib.makeOverridable (
+  attrs:
+  let
+    attrs' = if lib.isFunction attrs then attrs else _: attrs;
+    attrsFixed = lib.fix attrs';
+    stdenv' =
+      if attrsFixed.noCC or false then
+        stdenvNoCC
+      else if attrsFixed.noBootstrap or false then
+        stdenv
+      else
+        bootstrapStdenv;
+  in
+  stdenv'.mkDerivation (
+    lib.flip lib.extends attrs' (
+      self: super:
+      assert super ? releaseName;
+      let
+        inherit (super) releaseName;
+        info = (lib.importJSON ./versions.json).${releaseName};
```

Comment:
`lib.importJSON` can’t be at the top of the file because it needs to access `lib`. Is it okay to inline the definition of `lib.importJSON`? It’s trivial, but is it actually a worthwhile optimization?

Reply:
I’ve been pulling these JSON files up to the top using `builtins.fromJSON (builtins.readFile <filename)`. There were slight improvements in memory usage.